src/rgw/rgw_data_sync.cc

```diff
@@ -2286,33 +2288,41 @@ int RGWBucketShardFullSyncCR::operate()
           while (collect(&ret, lease_stack)) {
             if (ret < 0) {
               ldout(sync_env->cct, 0) << "ERROR: a sync operation returned error" << dendl;
+              sync_status = ret;
```

Comment:
`collect()` never returns true, so this block won't execute. removing the `while()` should work. same with `RGWBucketShardIncrementalSyncCR` below